File: leetcode/lc-us/cpp/131.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    vector<string> cur;
    vector<vector<string>> res;

    void dfs(string &s, int start) {
        if (start == s.length()) {
            res.push_back(cur);
            return;
        }
        for (int i = start; i < s.length(); i++) {
            if (!is_palindrome(s, start, i)) continue;
            cur.push_back(s.substr(start, i - start + 1));
            dfs(s, i + 1);
            cur.pop_back();
        }
    }

    bool is_palindrome(string &s, int low, int high) {
        while (low < high) {
            if (s[low] != s[high]) return false;
            low++;
            high--;
        }
        return true;
    }

public:
    vector<vector<string>> partition(string s) {
        dfs(s, 0);
        return res;
    }
};

File: leetcode/lc-us/py/131.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        path, res = [], []

        def valid(i: int, j: int) -> bool:
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True

        def dfs(u: int, pre: int, tot: int) -> None:
            if u == len(s):
                if tot == len(s): res.append(path[:])
                return
            dfs(u + 1, pre, tot)
            if valid(pre, u):
                path.append(s[pre: u + 1])
                dfs(u + 1, u + 1, tot + u - pre + 1)
                path.pop()

        dfs(0, 0, 0)
        return res
